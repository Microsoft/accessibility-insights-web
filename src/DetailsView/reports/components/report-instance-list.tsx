// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

export interface ReportInstanceListProps {
    nodeResults: AxeNodeResult[];
}

export class ReportInstanceList extends React.Component<ReportInstanceListProps> {
    public render(): JSX.Element {
        return (
            <div className="report-instances">
                <p>Failed elements:</p>
                <table className="report-instance-table">
                    <tbody>{this.renderInstances()}</tbody>
                </table>
            </div>
        );
    }

    private renderInstances(): JSX.Element[] {
        const rows: JSX.Element[] = [];
        this.props.nodeResults.forEach((nodeResult, index) => {
            rows.push(
                <tr className="path-row" key={`path-row-${index}`}>
                    <td className="label">Path</td>
                    <td className="content">{nodeResult.target}</td>
                </tr>,
            );
            rows.push(
                <tr className="snippet-row" key={`snippet-row-${index}`}>
                    <td className="label">Snippet</td>
                    <td className="content">{nodeResult.html}</td>
                </tr>,
            );
        });
        return rows;
    }
}
