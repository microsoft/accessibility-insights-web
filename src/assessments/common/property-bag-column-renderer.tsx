// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { InstanceTableRow } from 'assessments/types/instance-table-data';
import { ColumnValueBag } from 'common/types/property-bag/column-value-bag';
import { PropertyBagColumnRendererConfig } from 'common/types/property-bag/property-bag-column-renderer-config';
import { isEmpty } from 'lodash';
import * as React from 'react';

import { DictionaryStringTo } from 'types/common-types';

export function propertyBagColumnRenderer<TPropertyBag extends ColumnValueBag>(
    item: InstanceTableRow<TPropertyBag>,
    configs: PropertyBagColumnRendererConfig<TPropertyBag>[],
): JSX.Element {
    const mapper = (config: PropertyBagColumnRendererConfig<TPropertyBag>, index: number) => {
        if (item.instance.propertyBag == null) {
            return null;
        }
        const value = item.instance.propertyBag[config.propertyName];
        if (value == null && config.defaultValue == null) {
            return null;
        }

        return render(config, value, index);
    };

    const render = (
        config: PropertyBagColumnRendererConfig<TPropertyBag>,
        value: any,
        index: number,
    ) => {
        return (
            <div key={`property-${index}`} className="property-bag-div">
                <span className="display-name">{`${config.displayName}: `}</span>
                {renderValue(config, value)}
            </div>
        );
    };

    const renderValue = (config: PropertyBagColumnRendererConfig<TPropertyBag>, value: any) => {
        if (config.expand) {
            return renderProperties(config, value);
        }

        return <React.Fragment>{value || config.defaultValue}</React.Fragment>;
    };

    const renderProperties = (
        config: PropertyBagColumnRendererConfig<TPropertyBag>,
        propertyMap: DictionaryStringTo<string>,
    ) => {
        if (isEmpty(propertyMap)) {
            return <React.Fragment>{config.defaultValue.toString()}</React.Fragment>;
        }

        return Object.keys(propertyMap).map(key => {
            return (
                <div key={key} className="expanded-property-div">
                    {renderInnerKeyValue(key, propertyMap[key])}
                </div>
            );
        });
    };

    const renderInnerKeyValue = (key: string, value: string) => {
        if (value === null) {
            return <span className="display-name">{`${key}`}</span>;
        } else {
            return (
                <React.Fragment>
                    <span className="display-name">{`${key}: `}</span>
                    {value}
                </React.Fragment>
            );
        }
    };

    return <div className="property-bag-container">{configs.map(mapper)}</div>;
}
