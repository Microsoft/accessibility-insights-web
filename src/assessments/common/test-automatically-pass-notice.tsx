// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as React from 'react';
import * as Markup from '../markup';

const TestAutomaticallyPassedNotice = (
    <p>
        <Markup.Emphasis>
            Note: If no matching/failing instances are found, this requirement will automatically be marked as pass.
        </Markup.Emphasis>
    </p>
);

export default TestAutomaticallyPassedNotice;
