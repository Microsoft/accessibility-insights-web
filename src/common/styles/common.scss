// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
@import 'high-contrast-override.scss';
@import 'colors.scss';
@import 'fonts.scss';

$focus-ring-color: $neutral-100;

@mixin set-focus-ring-styles($property: 'border') {
    #{$property}: 1px solid $focus-ring-color;
    @include ms-high-contrast-override {
        #{$property}: 1px dashed windowtext !important;
    }
}

.ms-Fabric.is-focusVisible .ms-Nav-group .ms-nav-linkButton:focus::after {
    @include set-focus-ring-styles('border');
}

button::after {
    content: '';
    position: absolute;
}

.ms-Button.ms-Button--primary:not(.is-disabled) {
    background: $communication-primary;

    &:hover {
        background: $communication-shade-10;
    }
}

.ms-fontColor-neutralPrimary,
.ms-fontColor-neutralPrimary--hover:hover {
    color: $neutral-100 !important;
}

.ms-Fabric {
    color: $neutral-100;
}

.generic-toggle-component {
    margin-bottom: 4vh;
    .toggle-container {
        margin-bottom: 1vh;
        .toggle-name {
            display: inline-block;
            font-size: 17px;
            font-weight: bold;
        }

        .toggle {
            float: right;
        }
    }
}

.high-contrast-theme {
    .insights-link {
        &:hover {
            text-decoration: underline;
        }
    }
    .is-selected {
        .status-icon {
            color: $high-contrast-icon-color !important;
        }
    }
}

.insights-link {
    // marking important to override office fabric style
    color: $link !important;
    text-decoration: none;

    &:hover {
        color: $link-hover !important;
        text-decoration: underline;
    }
}

.ms-Spinner {
    .ms-Spinner-circle {
        border-top-color: $spinner-text;
    }

    .ms-Spinner-label {
        color: $spinner-text;
    }
}

$normalTitleFontWeight: 300 !default;

$pivotItemLeftBorderWidth: 3px !default;
$pivotItemBorderStyle: solid !default;

$detailsViewHeaderBarHeight: 40px !default;
$detailsViewNavPivotsHeight: 41px !default;
$detailsViewCommandBarHeight: 40px !default;
$detailsViewCommandBarBorderHeight: 1px !default;
$detailsViewLeftNavWidth: 200px !default;
$detailsViewTotalHeaderHeight: (#{$detailsViewCommandBarBorderHeight} + #{$detailsViewCommandBarHeight} + #{$detailsViewHeaderBarHeight});

$fastPassRightPanelMarginRight: 24px !default;
$fastPassRightPanelMarginLeft: 24px !default;
$fastPassRightPanelMarginTop: 24px !default;

@mixin ellipsedText {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
}
