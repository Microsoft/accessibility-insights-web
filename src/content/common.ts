// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as react from 'react';

import { link } from './link';
import { ContentCreator } from '../views/content/content-page';
import { GuidanceTitle as gt } from '../views/content/guidance-title';

export const create = ContentCreator(link);
export const React = react;
export const GuidanceTitle = gt;
