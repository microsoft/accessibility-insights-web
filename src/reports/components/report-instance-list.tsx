// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { NamedFC } from 'common/react/named-fc';
import {
    BagOf,
    isScalarColumnValue,
    ScalarColumnValue,
} from 'common/types/property-bag/column-value-bag';
import { flatten, toPairs } from 'lodash';
import * as React from 'react';

import { InstanceReportModel } from '../assessment-report-model';

export interface ReportInstanceListProps {
    instances: InstanceReportModel[];
}

export const ReportInstanceList = NamedFC<ReportInstanceListProps>('ReportInstanceList', props => {
    type Index = number | string;
    type Row = React.DetailedHTMLProps<
        React.HTMLAttributes<HTMLTableRowElement>,
        HTMLTableRowElement
    >;
    type Cell = React.DetailedHTMLProps<
        React.HTMLAttributes<HTMLTableCellElement>,
        HTMLTableCellElement
    >;

    return <div>{renderInstances()}</div>;

    function renderInstances(): JSX.Element[] {
        return props.instances.map((instance, index) => {
            return (
                <table className="instance-details" key={`instance-row-${index}`}>
                    <tbody>{renderInstanceRows(instance).toString()}</tbody>
                </table>
            );
        });
    }

    function renderInstanceRows(instance: InstanceReportModel): Row[] {
        const rowSets = instance.props.map(({ key, value }, index) =>
            isScalarColumnValue(value)
                ? [renderScalarRow(key, value, index)]
                : renderPropertyBagRows(key, value, index),
        );

        return flatten(rowSets);
    }

    function renderPropertyBagRows(
        key: string,
        value: BagOf<ScalarColumnValue>,
        index: Index,
    ): Row[] {
        const headerRow = renderPropertyBagHeaderRow(key);
        const entryRows = renderProperyBagEntryRows(value, index);
        return [headerRow, ...entryRows];
    }

    function renderPropertyBagHeaderRow(key: string): Row {
        return renderRow(
            [
                <th className={`instance-subsection-header`} colSpan={2} key={key}>
                    {key + ':'}
                </th>,
            ],
            key,
        );
    }

    function renderProperyBagEntryRows(bag: BagOf<ScalarColumnValue>, outerIndex: Index): Row[] {
        return toPairs(bag).map(([key, value], index) =>
            renderScalarRow(key, value, `${outerIndex}-${index}`, 'instance-key-indented'),
        );
    }

    function renderScalarRow(
        key: string,
        value: ScalarColumnValue,
        index: Index,
        keyClassName: string = 'instance-key',
    ): Row {
        return renderRow(
            [
                <th className={keyClassName} key={key}>
                    {key}
                </th>,
                <td className={'instance-value'} key={`${key}-${index}`}>
                    {value ? value.toString() : '-'}
                </td>,
            ],
            index,
        );
    }

    function renderRow(cells: Cell[], index: Index): Row {
        return (
            <tr className="instance-pair-details" key={`instance-pair-row-${index}`}>
                {cells.toString()}
            </tr>
        );
    }
});
