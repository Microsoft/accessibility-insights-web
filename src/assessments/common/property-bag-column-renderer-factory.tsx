// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ColumnValueBag } from '../../common/types/property-bag/column-value-bag';
import { IAssessmentInstanceRowData } from '../../DetailsView/components/assessment-instance-table';
import { propertyBagColumnRenderer, PropertyBagColumnRendererConfig } from './property-bag-column-renderer';

export class PropertyBagColumnRendererFactory {
    public static get<TPropertyBag extends ColumnValueBag>(
        configs: PropertyBagColumnRendererConfig<TPropertyBag>[],
    ): (item: IAssessmentInstanceRowData<TPropertyBag>) => JSX.Element {
        return item => {
            return propertyBagColumnRenderer(item, configs);
        };
    }
}
