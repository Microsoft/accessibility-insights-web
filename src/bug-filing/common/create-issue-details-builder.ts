// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { EnvironmentInfo } from '../../common/environment-info-provider';
import { CreateIssueDetailsTextData } from '../../common/types/create-issue-details-text-data';
import { title } from '../../content/strings/application';
import { IssueDetailsBuilder } from './issue-details-builder';
import { MarkupFactory } from './markup-factory';

export const createIssueDetailsBuilder = (markup: MarkupFactory): IssueDetailsBuilder => {
    const getter = (environmentInfo: EnvironmentInfo, data: CreateIssueDetailsTextData): string => {
        const result = data.ruleResult;

        const text = [
            `${markup.bold('Issue')}: ${markup.snippet(result.help)} (${markup.link(result.helpUrl, result.ruleId)})`,
            markup.sectionSeparator(),
            `${markup.bold('Target application')}: ${markup.link(data.pageUrl, data.pageTitle)}`,
            markup.sectionSeparator(),
            `${markup.bold('Element path')}: ${data.ruleResult.selector}`,
            markup.sectionSeparator(),
            `${markup.bold('Snippet')}:`,
            markup.snippet(result.snippet),
            markup.sectionSeparator(),
            `${markup.bold('How to fix')}:`,
            markup.newLine(),
            markup.howToFixSection(result.failureSummary),
            markup.sectionSeparator(),
            `${markup.bold('Environment')}:`,
            environmentInfo.browserSpec,
            markup.sectionSeparator(),
            `This accessibility issue was found using ${title} ` +
                `${environmentInfo.extensionVersion} (axe-core ${environmentInfo.axeCoreVersion}), ` +
                'a tool that helps find and fix accessibility issues. Get more information & download ' +
                `this tool at ${markup.link('http://aka.ms/AccessibilityInsights')}.`,
        ].join('\n');

        return text;
    };

    return getter;
};
