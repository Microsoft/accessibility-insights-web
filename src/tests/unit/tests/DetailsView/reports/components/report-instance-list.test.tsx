// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { ReportInstanceList, ReportInstanceListProps } from '../../../../../../DetailsView/reports/components/report-instance-list';
import { shallowRender } from '../../../../common/shallow-render';

describe('ReportInstanceListTest', () => {
    test('render 0 instances', () => {
        const props: ReportInstanceListProps = {
            nodeResults: [],
        };

        expect(shallowRender(<ReportInstanceList {...props} />)).toMatchSnapshot();
    });

    test('render 2 instances', () => {
        const props: ReportInstanceListProps = {
            nodeResults: [
                {
                    target: 'target-1',
                    html: '<html-1/>',
                } as any,
                {
                    target: 'target-2',
                    html: '<html-2/>',
                } as any,
            ],
        };

        expect(shallowRender(<ReportInstanceList {...props} />)).toMatchSnapshot();
    });
});
