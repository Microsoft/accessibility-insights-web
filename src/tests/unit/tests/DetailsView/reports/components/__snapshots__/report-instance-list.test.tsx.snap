// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportInstanceListTest render 0 instances 1`] = `
<div
  className="report-instances"
>
  <p>
    Failed elements:
  </p>
  <table
    className="report-instance-table"
  >
    <tbody />
  </table>
</div>
`;

exports[`ReportInstanceListTest render 2 instances 1`] = `
<div
  className="report-instances"
>
  <p>
    Failed elements:
  </p>
  <table
    className="report-instance-table"
  >
    <tbody>
      <tr
        className="path-row"
      >
        <td
          className="label"
        >
          Path
        </td>
        <td
          className="content"
        >
          target-1
        </td>
      </tr>
      <tr
        className="snippet-row"
      >
        <td
          className="label"
        >
          Snippet
        </td>
        <td
          className="content"
        >
          &lt;html-1/&gt;
        </td>
      </tr>
      <tr
        className="path-row"
      >
        <td
          className="label"
        >
          Path
        </td>
        <td
          className="content"
        >
          target-2
        </td>
      </tr>
      <tr
        className="snippet-row"
      >
        <td
          className="label"
        >
          Snippet
        </td>
        <td
          className="content"
        >
          &lt;html-2/&gt;
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
